runtime/src/main/java/org/corfudb/runtime/collections/streaming/DeltaStream.java

```diff
@@ -0,0 +1,184 @@
+import org.corfudb.runtime.exceptions.TrimmedException;
+import org.corfudb.runtime.view.Address;
+import org.corfudb.runtime.view.AddressSpaceView;
+import org.corfudb.runtime.view.ReadOptions;
+import org.corfudb.runtime.view.stream.StreamAddressSpace;
+
+import java.time.Duration;
+import java.util.UUID;
+
+/**
+ * Unlike the {@link org.corfudb.runtime.view.stream.IStreamView} this stream implementation only tracks updates
+ * that haven't been checkpointed.
+ */
+@Slf4j
+public class DeltaStream {
+
+    private final AddressSpaceView addressSpaceView;
+
+    @Getter
+    private final UUID streamId;
+
+    /**
+     * Last Address consumed/read by this stream
+     */
+    private volatile long lastAddressRead;
+
+    /**
+     * Max seen address for this stream from the last refresh/sync
+     */
+    @Getter
+    private volatile long maxAddressSeen;
+
+    /**
+     * The last trim mark that has been observed
+     */
+    private volatile long trimMark;
+
+    /**
+     * The size of this stream's buffer (i.e., pending addresses to read)
+     */
+    private final int bufferSize;
+
+    /**
+     * A circular buffer that holds the queued address to read
+     */
+    private final Buffer addressesToRead;
+
+    private final ReadOptions options = ReadOptions
+            .builder()
+            .clientCacheable(false)
+            .ignoreTrim(false)
+            .waitForHole(true)
+            .serverCacheable(false)
+            .build();
+
+    public DeltaStream(AddressSpaceView addressSpaceView, UUID streamId, long lastAddressRead, int bufferSize) {
+        this.addressSpaceView = addressSpaceView;
+        Preconditions.checkArgument(Address.NON_ADDRESS == -1);
+        Preconditions.checkArgument(lastAddressRead >= Address.NON_ADDRESS,
+                "lastAddressRead %s must be >= %s", lastAddressRead, Address.NON_ADDRESS);
+        Preconditions.checkArgument(bufferSize > 0, "buffer size %s must be > 0", bufferSize);
+        this.streamId = streamId;
+        this.bufferSize = bufferSize;
+        this.addressesToRead = BufferUtils.synchronizedBuffer(new BoundedFifoBuffer(this.bufferSize));
+        this.lastAddressRead = lastAddressRead;
+        this.maxAddressSeen = this.lastAddressRead;
+        this.trimMark = Address.NON_ADDRESS;
+    }
+
+    /**
+     * Check if a the last address read was trimmed, since the last address has to always be
+     * greater than the trim mark, if this condition is trim, then a trim has created a gap
+     * that hasn't be synced and cannot be recovered.
+     */
+    private boolean trimGap() {
+        return trimMark != Address.NON_ADDRESS && lastAddressRead < trimMark;
+    }
+
+    //max has to always be greater than last address read
+    // called by the scheduler thread
+    public int availableSpace() {
+        int availableSpace = this.bufferSize - addressesToRead.size();
+        Preconditions.checkState(availableSpace >= 0);
```

Comment:
That's always unexpected when you execute a methods like `size()`, `availableSpace()`, `isEmpty()` and the method throws an Exception. I'd say that's a bad practice

Reply:
The exception is not meant to be part of the API, this is more of an **invariant**. If this exception is thrown, then this signals that the internal state of a DeltaStream has been corrupted. Explicitly failing makes it easier to detect concurrency issues easier.

Reply:
Then you should move that check into the class constructor, it prevents you from creating a corrupted instance of DeltaStream? Why you need to check each time

Reply:
`addressesToRead.size()` changes as the worker thread consumes items from the buffer therefore checking it once in the constructor is not enough.

Reply:
> Explicitly failing makes it easier to detect concurrency issues easier.

This If corruption happens because of multi-threading, at construction time there is only one thread, so moving it in the constructor doesn't capture what im checking for here.